src/pipelines/sentence_embeddings/layers.rs

```diff
@@ -0,0 +1,134 @@
+pub struct PoolingConfig {
+    pub word_embedding_dimension: i64,
+    pub pooling_mode_cls_token: bool,
+    pub pooling_mode_max_tokens: bool,
+    pub pooling_mode_mean_tokens: bool,
+    pub pooling_mode_mean_sqrt_len_tokens: bool,
+}
+
+impl Config for PoolingConfig {}
+
+pub struct Pooling {
+    conf: PoolingConfig,
+}
+
+impl Pooling {
+    pub fn new(conf: PoolingConfig) -> Pooling {
+        Pooling { conf }
+    }
+
+    pub fn forward(&self, mut token_embeddings: Tensor, attention_mask: &Tensor) -> Tensor {
+        let mut output_vectors = Vec::new();
+
+        if self.conf.pooling_mode_cls_token {
+            let cls_token = token_embeddings.select(1, 0); // Take first token by default
+            output_vectors.push(cls_token);
+        }
+
+        if self.conf.pooling_mode_max_tokens {
+            let input_mask_expanded = attention_mask.unsqueeze(-1).expand_as(&token_embeddings);
+            // Set padding tokens to large negative value
+            token_embeddings = token_embeddings.masked_fill_(&input_mask_expanded.eq(0), -1e9);
+            let max_over_time = token_embeddings.max_dim(1, true).0;
+            output_vectors.push(max_over_time);
+        }
+
+        if self.conf.pooling_mode_mean_tokens || self.conf.pooling_mode_mean_sqrt_len_tokens {
+            let input_mask_expanded = attention_mask.unsqueeze(-1).expand_as(&token_embeddings);
+            let sum_embeddings =
+                (token_embeddings * &input_mask_expanded).sum_dim_intlist(&[1], false, Kind::Float);
+
+            let sum_mask = input_mask_expanded.sum_dim_intlist(&[1], false, Kind::Float);
+            let sum_mask = sum_mask.clamp_min(10e-9);
+
+            if self.conf.pooling_mode_mean_tokens {
+                output_vectors.push(&sum_embeddings / &sum_mask);
+            }
+            if self.conf.pooling_mode_mean_sqrt_len_tokens {
+                output_vectors.push(sum_embeddings / sum_mask.sqrt());
+            }
+        }
+
+        Tensor::cat(&output_vectors, 1)
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct DenseConfig {
+    pub in_features: i64,
+    pub out_features: i64,
+    pub bias: bool,
+    #[serde(deserialize_with = "last_part")]
+    pub activation_function: Activation,
+}
+
+impl Config for DenseConfig {}
+
+/// Split the given string on `.` and try to construct an `Activation` from the last part
+fn last_part<'de, D>(deserializer: D) -> Result<Activation, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let activation = String::deserialize(deserializer)?;
+    activation
+        .split('.')
+        .last()
+        .map(|s| serde_json::from_value(serde_json::Value::String(s.to_lowercase())))
+        .transpose()
+        .map_err(de::Error::custom)?
+        .ok_or_else(|| format!("Invalid Activation: {}", activation))
+        .map_err(de::Error::custom)
+}
+
+pub struct Dense {
+    linear: nn::Linear,
+    activation: TensorFunction,
+    _var_store: nn::VarStore,
```

Comment:
Are `_var_store` and `_conf` used at a later point?

Reply:
Regarding `_var_store`, since some weights are loaded from another file specifically for this layer (in the original Python implementation), I think it's good to keep it as part of the struct in order to have a clean lifecycle (i.e. the layer is dropped => this sub VarStore is dropped too).

As for `_conf` I kept it in case it would be used in `forward(...)`, or for debugging purposes. But I can remove it if you prefer.

Reply:
I see I understand now - one alternative would be to manage the `var_store` from the SentenceEmbeddingsModel`, as is done for the transformer. This is also fine. I would however remove the `_config` object, it's value can be observed from `SentenceEmbeddingsModel` when debugging anyway.

Reply:
Done